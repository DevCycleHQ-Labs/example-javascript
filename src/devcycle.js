import { initializeDevCycle } from "@devcycle/js-client-sdk";
import { users } from "./users";

// Create DevCycle client and set up event listeners
export const setUpDevCycle = () => {
    if (!process.env.DEVCYCLE_CLIENT_SDK_KEY) {
        alert('Set your DEVCYCLE_CLIENT_SDK_KEY environment variable to use the DevCycle JavaScript SDK.')
    }

    // Initialize the DevCycle client with your SDK key and user
    const devcycleOptions = { logLevel: "debug" };
    const devcycleClient = initializeDevCycle(
        process.env.DEVCYCLE_CLIENT_SDK_KEY,
        users[0], // identifying initial user as user-1
        devcycleOptions
    );

    setUpIdentifyDropdown(devcycleClient);
    document.getElementById("reset").onclick = () => {
        devcycleClient.resetUser().then(() => updateUI(devcycleClient));
        document.getElementById("identify").options[3].selected = true;
    };

    // Update the app when DevCycle receives the first user config
    devcycleClient.onClientInitialized().then(({ config }) => {
        updateUI(devcycleClient);
    });

    // Update the app when DevCycle receives realtime updates from the dashboard
    devcycleClient.subscribe('configUpdated', () => {
        updateUI(devcycleClient);
    });
}

const setUpIdentifyDropdown = (devcycleClient) => {
    const identifyDropdown = document.getElementById("identify");

    identifyDropdown.innerHTML = `
        ${users.map(user => (
         `<option key='${user.user_id}' value='${JSON.stringify(user)}'>${user.name}</option>`
        ))}
        <option key="anonymous" value='{}'>Anonymous User</option>
    `

    identifyDropdown.onchange = (event) => {
        devcycleClient.identifyUser(JSON.parse(event.target.value)).then(() => updateUI(devcycleClient));
    }
}

// Use DevCycle variables to control the UI 
const updateUI = (devcycleClient) => {
    const greeting = devcycleClient.variableValue('togglebot-greeting', 'Hello world!')
    const shouldWink = devcycleClient.variableValue('togglebot-wink', false)
    const spinSpeed = devcycleClient.variableValue('togglebot-speed', 'off')
    
    document.getElementById("greeting").innerHTML = greeting;
    const togglebot = document.getElementById("togglebot");
    togglebot.src = shouldWink ? "./images//togglebot-wink.png" : "./images/togglebot.png";

    togglebot.classList.remove("spin-slow", "spin-fast", "spin-super-fast", "spin-off");
    togglebot.classList.add(`spin-${spinSpeed}`);
};
