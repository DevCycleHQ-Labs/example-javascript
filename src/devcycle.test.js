import { fireEvent, waitFor } from '@testing-library/dom';
import fs from 'fs';
import path from 'path';
import '@testing-library/jest-dom';
import { setUpDevCycle } from './devcycle.js';
import { users } from './users.js';

process.env.DEVCYCLE_CLIENT_SDK_KEY = 'mocked-sdk-key';

const mockAllFeatures = jest.fn(() => ({}))
const mockDevCycleClient = {
    identifyUser: jest.fn(() => Promise.resolve(true)),
    resetUser: jest.fn(() => Promise.resolve(true)),
    onClientInitialized: jest.fn(() => Promise.resolve(true)),
    variableValue: jest.fn(),
    subscribe: jest.fn(),
    allFeatures: mockAllFeatures
}

const mockVariableValue = (variable, value) => {
    mockDevCycleClient.variableValue.mockImplementation((variableKey, defaultValue) => {
        return variable === variableKey ? value : defaultValue
    })
}

jest.mock('@devcycle/js-client-sdk', () => ({
    initializeDevCycle: () => mockDevCycleClient
}));

const html = fs.readFileSync(path.resolve(__dirname, '../public/index.html'), 'utf8');

describe('DevCycle client initialization', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        document.documentElement.innerHTML = html.toString();
    });

    test('App initializes DevCycle client correctly', async () => {
        setUpDevCycle();
        await waitFor(() => expect(mockDevCycleClient.onClientInitialized).toHaveBeenCalled());
    });
})

describe('Togglebot speed', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        document.documentElement.innerHTML = html.toString();
    });

    test('Togglebot handles speed "off"', async () => {
        mockVariableValue('togglebot-speed', 'off');
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('togglebot-message')).toHaveTextContent("Hello! Nice to meet you.");
        });
    })

    test('Togglebot handles speed "slow"', async () => {
        mockVariableValue('togglebot-speed', 'slow');
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('togglebot-message')).toHaveTextContent("Awesome, look at you go!");
        });
    })

    test('Togglebot handles speed "fast"', async () => {
        mockVariableValue('togglebot-speed', 'fast');
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('togglebot-message')).toHaveTextContent("This is fun!");
        });
    })

    test('Togglebot handles speed "off-axis"', async () => {
        mockVariableValue('togglebot-speed', 'off-axis');
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('togglebot-message')).toHaveTextContent("...I’m gonna be sick...");
        });
    })


    test('Togglebot handles speed "surprise"', async () => {
        mockVariableValue('togglebot-speed', 'surprise');
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('togglebot-message')).toHaveTextContent("What the unicorn?");
        });
    })
})

describe('App instructions', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        document.documentElement.innerHTML = html.toString();
    });

    test('App handles default variation', async () => {
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('variation-name')).toHaveTextContent('Default');
            expect(document.getElementById('instructions-header')).toHaveTextContent('Welcome to DevCycle\'s example app');
        });
    })

    test('App handles base variation', async () => {
        mockAllFeatures.mockImplementation(() => ({
            'hello-togglebot': {
                variationKey: 'variation-base',
                variationName: 'Base',
            }
        }));
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('variation-name')).toHaveTextContent('Base');
            expect(document.getElementById('instructions-header')).toHaveTextContent('Welcome to DevCycle\'s example app');
        });
    })

    test('App handles wink variation', async () => {
        mockAllFeatures.mockImplementation(() => ({
            'hello-togglebot': {
                variationKey: 'variation-wink',
                variationName: 'Wink',
            }
        }));
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('variation-name')).toHaveTextContent('Wink');
            expect(document.getElementById('instructions-header')).toHaveTextContent('Great! You\'ve taken the first step in exploring DevCycle.');
        });
    })


    test('App handles custom variation', async () => {
        mockAllFeatures.mockImplementation(() => ({
            'hello-togglebot': {
                variationKey: 'variation-custom',
                variationName: 'Custom',
            }
        }));
        setUpDevCycle();
        await waitFor(() => {
            expect(document.getElementById('variation-name')).toHaveTextContent('Custom');
            expect(document.getElementById('instructions-header')).toHaveTextContent('You\’re getting the hang of things.');
        });
    })

})

