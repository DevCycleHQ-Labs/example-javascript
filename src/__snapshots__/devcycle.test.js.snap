// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App content App description is updated for value "default" 1`] = `"Welcome to DevCycle's example app."`;

exports[`App content App description is updated for value "default" 2`] = `
"
    
        If you got to the example app on your own, follow our README guide to create the Feature and Variables you need in DevCycle to control this app.
    
"
`;

exports[`App content App description is updated for value "step-1" 1`] = `"Welcome to DevCycle's example app."`;

exports[`App content App description is updated for value "step-1" 2`] = `
"
    
     If you got here through the onboarding flow, just follow the instructions to change and create new Variations and see how the app reacts to new Variable values.
    
"
`;

exports[`App content App description is updated for value "step-2" 1`] = `"Great! You've taken the first step in exploring DevCycle."`;

exports[`App content App description is updated for value "step-2" 2`] = `
"
    
        You've successfully toggled your very first variation. You are now serving a different value to your users and you can see how the example app has reacted to this change.
    
    
        Next, go ahead and create a whole new variation to see what else is possible in this app.
    
"
`;

exports[`App content App description is updated for value "step-3" 1`] = `"Welcome to DevCycle's example app."`;

exports[`App content App description is updated for value "step-3" 2`] = `
"
    
        If you got to the example app on your own, follow our README guide to create the Feature and Variables you need in DevCycle to control this app.
    
"
`;
